paasta_tools/frameworks/native_scheduler.py

```diff
@@ -125,22 +140,35 @@ def launch_tasks_for_offers(self, driver, offers):
         launched_tasks = []
 
         for offer in offers:
-            tasks = self.tasks_for_offer(driver, offer)
-            if tasks:
-                for task in tasks:
-                    self.tasks_with_flags.setdefault(
-                        task.task_id.value,
-                        MesosTaskParameters(
-                            health=None, mesos_task_state=TASK_STAGING))
-
-                operation = mesos_pb2.Offer.Operation()
-                operation.type = mesos_pb2.Offer.Operation.LAUNCH
-                operation.launch.task_infos.extend(tasks)
-                driver.acceptOffers([offer.id], [operation])
-
-                launched_tasks.extend(tasks)
-            else:
-                driver.declineOffer(offer.id)
+            self.constraint_state_lock.acquire()
```

Comment:
im fairly sure you can use the lock as a context manager - you can forget about explicitly dropping the lock that way.

```python
with self.constraint_state_lock.acquire():
     ...
```